File: benchmarks/livermore_canonical/k08_adi.f
C     Kernel 8 -- ADI integration (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical: U1(5,101,2), U2(5,101,2), U3(5,101,2), DU1/2/3(101)
C
      PROGRAM K08
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION U1(5,101,2), U2(5,101,2), U3(5,101,2)
      DIMENSION DU1(101), DU2(101), DU3(101)
      DIMENSION SPACER(39)
C
C     Initialize SPACER and pull A11..A33, SIG
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 5 K = 1, 39
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          SPACER(K) = (BUZZ - FIZZ) * 0.1D0
    5 CONTINUE
      A11 = SPACER(1)
      A12 = SPACER(2)
      A13 = SPACER(3)
      A21 = SPACER(4)
      A22 = SPACER(5)
      A23 = SPACER(6)
      A31 = SPACER(7)
      A32 = SPACER(8)
      A33 = SPACER(9)
      SIG = SPACER(34)
C
C     Initialize U1 in column-major order: K3 outer, K2 mid, K1 inner
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 12 K3 = 1, 2
          DO 11 K2 = 1, 101
              DO 10 K1 = 1, 5
                  BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
                  FUZZ = -FUZZ
                  U1(K1,K2,K3) = (BUZZ - FIZZ) * 0.1D0
   10         CONTINUE
   11     CONTINUE
   12 CONTINUE
C
C     Initialize U2
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 22 K3 = 1, 2
          DO 21 K2 = 1, 101
              DO 20 K1 = 1, 5
                  BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
                  FUZZ = -FUZZ
                  U2(K1,K2,K3) = (BUZZ - FIZZ) * 0.1D0
   20         CONTINUE
   21     CONTINUE
   22 CONTINUE
C
C     Initialize U3
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 32 K3 = 1, 2
          DO 31 K2 = 1, 101
              DO 30 K1 = 1, 5
                  BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
                  FUZZ = -FUZZ
                  U3(K1,K2,K3) = (BUZZ - FIZZ) * 0.1D0
   30         CONTINUE
   31     CONTINUE
   32 CONTINUE
C
C     Kernel loop
C
      N = 100
      DO 200 REP = 1, 1000000
          NL1 = 1
          NL2 = 2
          FW = 2.0D0
          DO 110 KX = 2, 3
              DO 100 KY = 2, N
                  DU1(KY) = U1(KX,KY+1,NL1) - U1(KX,KY-1,NL1)
                  DU2(KY) = U2(KX,KY+1,NL1) - U2(KX,KY-1,NL1)
                  DU3(KY) = U3(KX,KY+1,NL1) - U3(KX,KY-1,NL1)
                  U1(KX,KY,NL2) = U1(KX,KY,NL1)
     &                + A11*DU1(KY) + A12*DU2(KY) + A13*DU3(KY)
     &                + SIG*(U1(KX+1,KY,NL1) - FW*U1(KX,KY,NL1)
     &                       + U1(KX-1,KY,NL1))
                  U2(KX,KY,NL2) = U2(KX,KY,NL1)
     &                + A21*DU1(KY) + A22*DU2(KY) + A23*DU3(KY)
     &                + SIG*(U2(KX+1,KY,NL1) - FW*U2(KX,KY,NL1)
     &                       + U2(KX-1,KY,NL1))
                  U3(KX,KY,NL2) = U3(KX,KY,NL1)
     &                + A31*DU1(KY) + A32*DU2(KY) + A33*DU3(KY)
     &                + SIG*(U3(KX+1,KY,NL1) - FW*U3(KX,KY,NL1)
     &                       + U3(KX-1,KY,NL1))
  100         CONTINUE
  110     CONTINUE
  200 CONTINUE
C
      WRITE(*,900) U1(2,2,2)
  900 FORMAT('u1(2,2,2) = ', E25.15)
      END

File: benchmarks/livermore_canonical/k08_adi.c
/* K8 -- ADI integration (canonical, 1-based indexing)
   Canonical: U1(5,101,2), U2(5,101,2), U3(5,101,2), DU1/2/3(101)
   Column-major: k1 stride 1, k2 stride 5, k3 stride 5*101=505 */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     100
#define NREPS 22600000

/* 1-based Fortran column-major: U(arr,k1,k2,k3) with dims (5,101,2) */
#define U(arr,k1,k2,k3) arr[((k3)-1)*505 + ((k2)-1)*5 + (k1)]

int main(void) {
    /* Max flat: (2-1)*505 + (101-1)*5 + 5 = 1010, need [1011] */
    double u1[1011], u2[1011], u3[1011];
    double du1[102], du2[102], du3[102];
    double spacer[40];

    signel(spacer+1, 39);
    double a11 = spacer[1], a12 = spacer[2], a13 = spacer[3];
    double a21 = spacer[4], a22 = spacer[5], a23 = spacer[6];
    double a31 = spacer[7], a32 = spacer[8], a33 = spacer[9];
    double sig = spacer[34];

    /* Fill 1010 elements starting at index 1 (column-major storage order) */
    signel(&u1[1], 1010);
    signel(&u2[1], 1010);
    signel(&u3[1], 1010);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 0; rep < NREPS; rep++) {
        long nl1 = 1, nl2 = 2;
        double fw = 2.0;
        for (long kx = 2; kx <= 3; kx++) {
            for (long ky = 2; ky <= N; ky++) {
                du1[ky] = U(u1,kx,ky+1,nl1) - U(u1,kx,ky-1,nl1);
                du2[ky] = U(u2,kx,ky+1,nl1) - U(u2,kx,ky-1,nl1);
                du3[ky] = U(u3,kx,ky+1,nl1) - U(u3,kx,ky-1,nl1);
                U(u1,kx,ky,nl2) = U(u1,kx,ky,nl1)
                    + a11*du1[ky] + a12*du2[ky] + a13*du3[ky]
                    + sig*(U(u1,kx+1,ky,nl1) - fw*U(u1,kx,ky,nl1)
                           + U(u1,kx-1,ky,nl1));
                U(u2,kx,ky,nl2) = U(u2,kx,ky,nl1)
                    + a21*du1[ky] + a22*du2[ky] + a23*du3[ky]
                    + sig*(U(u2,kx+1,ky,nl1) - fw*U(u2,kx,ky,nl1)
                           + U(u2,kx-1,ky,nl1));
                U(u3,kx,ky,nl2) = U(u3,kx,ky,nl1)
                    + a31*du1[ky] + a32*du2[ky] + a33*du3[ky]
                    + sig*(U(u3,kx+1,ky,nl1) - fw*U(u3,kx,ky,nl1)
                           + U(u3,kx-1,ky,nl1));
            }
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("u1(2,2,2) = %f\n", U(u1,2,2,2));
    return 0;
}
